File: src/lib/events.ts
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export interface ConfigChangedPayload {
  // empty
}

export interface SnapshotCreatedPayload {
  folderId: string;
  relativePath: string;
  commitSha: string;
  timestamp: string;
}

export interface WatcherErrorPayload {
  folderId: string;
  message: string;
}

export interface RestoreCompletedPayload {
  folderId: string;
  relativePath: string;
  restoredFromSha: string;
}

export function onConfigChanged(
  handler: (payload: ConfigChangedPayload) => void,
): Promise<UnlistenFn> {
  return listen<ConfigChangedPayload>("config-changed", (e) => {
    handler(e.payload);
  });
}

export function onSnapshotCreated(
  handler: (payload: SnapshotCreatedPayload) => void,
): Promise<UnlistenFn> {
  return listen<SnapshotCreatedPayload>("snapshot-created", (e) => {
    handler(e.payload);
  });
}

export function onWatcherError(
  handler: (payload: WatcherErrorPayload) => void,
): Promise<UnlistenFn> {
  return listen<WatcherErrorPayload>("watcher-error", (e) => {
    handler(e.payload);
  });
}

export function onRestoreCompleted(
  handler: (payload: RestoreCompletedPayload) => void,
): Promise<UnlistenFn> {
  return listen<RestoreCompletedPayload>("restore-completed", (e) => {
    handler(e.payload);
  });
}

File: src-tauri/src/events.rs
//! Backend → frontend event payloads.

use serde::Serialize;

/// Fired whenever config is mutated (watcher should reload).
#[derive(Clone, Serialize)]
pub struct ConfigChangedPayload {}

#[derive(Clone, Serialize)]
pub struct SnapshotCreatedPayload {
    pub folder_id: String,
    pub relative_path: String,
    pub commit_sha: String,
    pub timestamp: String,
}

#[derive(Clone, Serialize)]
pub struct WatcherErrorPayload {
    pub folder_id: String,
    pub message: String,
}

#[derive(Clone, Serialize)]
#[allow(dead_code)] // emitted from restore flow (step 11+)
pub struct RestoreCompletedPayload {
    pub folder_id: String,
    pub relative_path: String,
    pub restored_from_sha: String,
}
